src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java

```diff
@@ -228,16 +289,81 @@ public boolean shouldClearAll(@Nonnull Run<?,?> build) {
 
     }
 
-    @Restricted(DoNotUse.class)
+    /**
+     * Mixin interface for an {@link ArtifactManager} which supports specialized stash behavior as well.
+     */
+    @Restricted(Beta.class)
+    public interface StashAwareArtifactManager /* extends ArtifactManager */ {
```

Comment:
There is only one foreseen implementation of this beta SPI in Essentials: `JCloudsArtifactManager`. This in turn delegates to a small amount of provider-specific code to handle S3, etc.

Nor is baking protections into the API feasible. The specifics of JClouds semantics are interleaved with Remoting boundaries.

Modifying behavior of local artifact storage is out of scope. No one that I know of has complained about such issues with the current implementation, this PR does not touch it, and our policy moving forward will be to recommend external storage anyway.

Reply:
@jglick  Some points:

1. Whatever API you're expecting users to extend when writing new External Storage implementations needs to include hooks for JENKINS-50597 protections.  As we've already discussed a "one size fits all" solution isn't appropriate for handling network failures, and handling network failures is non-optional. We already know from durable-task-step and remoting failures that network failures come in many flavors, are common, and if we do not plan for them they can cause catastrophic stability problems.

2. If **this** isn't the API users are expected to implement/extend for external storage implementations, what is?  How do we ensure plugin authors base off that rather than this API? (Restrictions? Documentation?)  How do we ensure failures are appropriately propagated to the top level? I think IOException subtypes are generally appropriate for purposes of keeping API compatibility on current APIs, but would like the optional ability to inject a handler where practical.

> There is only one foreseen implementation of this beta SPI

This assumption raises a red flag, because if feels like we're getting tunnel vision about shipping one implementation and setting up for future problems.  I can totally understand the desire to just *release*, but if you want external storage to be broadly used, the APIs for it should be built for the future uses.

If you only want one API consumer for **this** API, it should be locked down as much as possible to prevent others and at the very least documented that "you should extend this other thing here, extending this is dangerous".  Otherwise the API should be built in a way that is appropriate to handle extension. 

Now, if the expected behavior is that plugin authors will *always* extend off JCloud APIs for external storage implementations, I have no problem with making that the case, but again we need to make that clear and do what we can to keep this locked-down -- and ensure again that failures are propagated.

> Nor is baking protections into the API feasible. The specifics of JClouds semantics are interleaved with Remoting boundaries.

That is going to need a proper explanation, I'm afraid -- on the surface it smells to me like we're letting an implementation detail dictate API design rather than trying to build for robustness and then figure out implementation.

> No one that I know of has complained about such issues with the current implementation

You mean, aside from general broken-ness with IO operations when an NFS server gets overloaded?  I don't think anybody is saying that's a hard requirement in this work, just that if our protections extend to that as a side-effect it won't be a problem and in fact may be beneficial.

> our policy moving forward will be to recommend external storage anyway

Which will require a rich ecosystem of implementations for different technologies, given how many different Jenkins deployment architectures there are.  Which again means that we need to build those base APIs to be reusable and safe for extension.